docs/standard/exceptions/best-practices-for-handling-and-using-exceptions-outline.md

```diff
@@ -0,0 +1,91 @@
+---
+title: [Best practices for handling and using exceptions]
+description: [Practical guide to handling and using exceptions.]
+author: [sam.wheat@outlook.com]
+ms.date: [12/10/2018]
+---
+# Best practices for handling and using exceptions
+
+@BillWagner For your review 
+
+1.	An approach to using exceptions and exception handling to your advantage.
+	1.	Start with reasonable expectations.
+		i.	Fault-tolerant is another way of saying that you have found a way to work around errors.
+		ii.	Robust recovery code adds significant complexity to your application.
+			1.	Adds additional code paths to troubleshoot when debugging
+			2.	Is sometimes useless since some exceptions are unrecoverable.
+	2.	Catching an exception is not itself a fix.
+		i.	A catch block offers you opportunity to gather information and begin recovery.
+		ii.	Catching exceptions incorrectly weakens your code.
+			1.	An exception often means your business logic is not doing its job.
+	3.	Preventing exceptions, not recovering from them, should be your first design objective.  

```

Comment:
I like this as a goal. But it's not practical. Any distributed system must be designed for exceptions that are not preventable: network requests may fail occasionally, timeouts may be exceeded, and so on.  

You're right to emphasize that you should develop code that enables business rule validation without generating exceptions. And that creates higher quality code. But other exceptions will occur because of the environment, hardware in IOT systems, and so on. 

There needs to be a section on how to manage those situations where the exception is not generated from a coding error, but from some condition outside of any programmer's control. Developers will want to know how to avoid corrupting state, how to retry, how to continue with only some features working.

Reply:
Yes this point needs clarification. The objective of being 100% bug free is an ideal but for all intents and purposes, not achievable. 
Even so, we should continue to design systems with the ongoing objective of being 100% bug free.  Being a first objective gives us room to add more objectives so perhaps I should be explicit and add something to the effect of  "Resilience should be your second goal".

> There needs to be a section on how to manage those situations where the exception is not generated from a coding error, but from some condition outside of any programmer's control.

Section 2.2 is devoted to this: "Use try/catch blocks to isolate code that may throw exceptions for reasons beyond your control."

> Developers will want to know how to avoid corrupting state, how to retry, how to continue with only some features working.

I'm not sure what to put here since this will be very specific to the application (also seems out of scope).

Reply:
@BillWagner updated per your comments

Reply:
@BillWagner   Any movement on this?

Reply:
@BillWagner  Any progress Bill?  It's not cool to ask me to work on this and then abandon it.

Reply:
@leaderanalytics I apologize. I missed seeing these notifications.  Looking now.